### ISSUE TITLE
Nine silent-failure defects in metrics, seeding, shuffling and checkpointing

### ISSUE THREAD

**Issue by mathnathan** [emelex-ai/BRIDGE#224]
## Summary

Nine defects, all present at `origin/main` (`5d7e8d9`), that share one failure mode: **they produce a
plausible result instead of an error.** Nothing crashes, nothing logs a warning, and the output looks
reasonable. Each one is independently reproducible and independently fixable.

They are grouped below by what they cost. Defect 1 corrupts every phonological metric BRIDGE reports.
Defects 2 to 5 silently break reproducibility and checkpointing. Defects 6 to 9 are ergonomic: they
force any downstream consumer to reach past the public API.

Line numbers are against `5d7e8d9`.

---

# 1. Every phonological metric counts padded positions

**Where:** `bridge/application/training/phon_metrics.py`, six sites: lines 8, 38, 56, 76, 102, 126.

**Now:**

```python
phon_valid_mask = phon_true != 2          # line 126
mask = true != 2                          # lines 38, 56, 76, 102
word[target != 2].all().int()             # line 8
```

**The problem:** `phon_true` is `phonology.phon_targets`, whose values are `{0, 1, phon_pad_id}`.
`phon_pad_id` is `phoneme_tokenizer.special_token_dims["[PAD]"]`, which is 35, not 2. The value 2
never occurs in the tensor, so `!= 2` is a tautology and the mask selects everything.

Verified against `5d7e8d9`:

```python
from bridge import BridgeTokenizer
enc = BridgeTokenizer().encode(["cat", "elephant", "read"])
t = enc.phonological.targets
sorted(t.unique().tolist())          # [0, 1, 35]
(t != 2).float().mean().item()       # 1.0  <- nothing is masked
```

**What it causes:** every metric in `calculate_phon_metrics` is computed over padded positions as
well as real ones. Because padding is trivially predictable, all of them are inflated, and the
inflation varies with how much padding a given batch happened to contain. Two runs with identical
model quality report different accuracy if their batches were padded differently. Feature accuracy,
phoneme-wise accuracy, word accuracy, euclidean distance, cosine distance and all three
closest-phoneme metrics are affected.

**Training itself is fine.** The loss uses `CrossEntropyLoss(ignore_index=35)`, which is correct.
This is a reporting defect, not an optimization one, which is exactly why it has survived: the models
train correctly and only the numbers describing them are wrong.

**Fix:** replace the literal `2` with the real pad id at all six sites. The pad id should be threaded
in rather than hardcoded again, since `VocabSpec.phon_pad_id` already carries it:

```python
def calculate_phon_metrics(logits, phonology, phon_reps, phon_pad_id: int) -> dict[str, float]:
    ...
    phon_valid_mask = phon_true != phon_pad_id
```

and the same substitution in the five helpers, each taking `phon_pad_id` as a parameter.

**Test:** encode a batch whose words have clearly different phoneme counts, so padding is guaranteed,
then assert the valid mask excludes it.

```python
def test_phon_metrics_mask_excludes_padding():
    enc = BridgeTokenizer().encode(["a", "elephant"])
    targets = enc.phonological.targets
    pad_id = BridgeTokenizer().phon_pad_id
    mask = targets != pad_id
    assert mask.float().mean().item() < 1.0, "mask must exclude padded positions"
    assert 2 not in targets.unique().tolist(), "2 is not a sentinel in this tensor"
```

---

# 2. `seed = 0` silently trains unseeded

**Where:** `bridge/domain/model/model.py:37`

**Now:**

```python
if self.model_config.seed:
    set_seed(seed=self.model_config.seed)
```

**The problem:** `0` is falsy, so a caller who passes `seed=0` gets no seeding at all. No warning is
logged. The model trains fine and is simply not reproducible.

**What it causes:** anyone building an ensemble with zero-indexed members (`for i in range(n)`) gets a
member 0 that is unseeded while members 1 through n-1 are seeded. Re-running the ensemble reproduces
every member except the first, and the discrepancy looks like ordinary training noise.

**Fix:**

```python
if self.model_config.seed is not None:
    set_seed(seed=self.model_config.seed)
```

**Test:**

```python
def test_seed_zero_is_honoured():
    cfg_a = ModelConfig(..., seed=0)
    a = Model(cfg_a)
    b = Model(ModelConfig(..., seed=0))
    for p, q in zip(a.parameters(), b.parameters()):
        assert torch.equal(p, q), "seed=0 must produce identical initialization"
```

This test fails today.

---

# 3. `TrainingPipeline` never shuffles the dataset

**Where:** `bridge/application/training/training_pipeline.py`, the epoch loop at line 422.
`BridgeDataset.shuffle()` exists at `bridge/domain/data/bridge_dataset.py:292` and is never called
from the pipeline. `grep -c shuffle training_pipeline.py` returns 0.

**The problem:** every epoch iterates the dataset in identical file order. There is no `DataLoader`
and no shuffle call, so example order is fixed for the entire run and identical across every model
trained on the same data.

**What it causes:** training order cannot be varied. Two models differing only in a seed intended to
control data order are trained on byte-identical sequences, so that source of variation contributes
exactly zero and nothing reports a problem. Any experiment treating data order as an independent
variable is silently measuring nothing.

It also means the model sees the data in whatever order the source file happened to have, which for
an alphabetically sorted lexicon means it trains on all the *a* words first.

**Fix:** shuffle the training partition between epochs, leaving the validation tail fixed. Note that
`BridgeDataset.shuffle` uses the global `random` module (`bridge_dataset.py:301`), so seeding must go
through `random.seed`:

```python
cut = int(len(self.dataset) * self.training_config.train_test_split)
for epoch in range(self.start_epoch, self.training_config.num_epochs):
    if self.training_config.shuffle_each_epoch:
        random.seed(self.training_config.seed * 10_000 + epoch)
        self.dataset.shuffle(cut)
    ...
```

with `shuffle_each_epoch: bool = Field(default=True)` added to `TrainingConfig`. Defaulting to True is
the right choice; defaulting to False preserves a bug for compatibility.

**Test:**

```python
def test_epochs_see_different_orders():
    orders = [list(pipeline.dataset.words) for _ in run_two_epochs(pipeline)]
    assert orders[0] != orders[1], "training order must vary between epochs"
```

---

# 4. Resume is dead code

**Where:** `bridge/application/training/training_pipeline.py:471-482`

**Now:**

```python
if "epoch" in checkpoint:
    if self.training_config.checkpoint_path and (...):
        self.start_epoch = checkpoint["epoch"] + 1
        self.logger.info(f"Resuming training from epoch {self.start_epoch}")
else:
    self.logger.warning("Checkpoint doesn't contain epoch information, starting from 0")
    self.start_epoch = 0
self.start_epoch = 0          # line 482: unconditional, overwrites the line above
```

**The problem:** line 482 runs unconditionally and discards whatever the resume branch computed. The
log line at 478 still prints "Resuming training from epoch 7", so the logs claim a resume that did not
happen.

**What it causes:** setting `checkpoint_path` loads pre-trained weights and then restarts the epoch
counter at 0. The run trains for the full `num_epochs` again starting from an already-trained model,
which is a different experiment from the one requested. The log actively misleads anyone checking.

**Fix:** delete line 482. Initialize `self.start_epoch = 0` at `__init__` (already done at line 58) and
let the resume branch be the only thing that changes it.

**Test:**

```python
def test_resume_starts_from_the_checkpoint_epoch():
    pipeline.save_model(epoch=3, run_name="x")
    fresh = TrainingPipeline(model, cfg_with_checkpoint_path, dataset, logger)
    assert fresh.start_epoch == 4
```

---

# 5. `save_model` accepts `run_name` and ignores it

**Where:** `bridge/application/training/training_pipeline.py:434-436`

**Now:**

```python
def save_model(self, epoch: int, run_name: str) -> None:
    if (epoch + 1) % self.training_config.save_every == 0:
        model_path = f"{self.training_config.model_artifacts_dir}/model_epoch_{epoch}.pth"
```

**The problem:** `run_name` is in the signature, is passed at the call site (line 431), and is never
used. The filename depends only on the epoch.

**What it causes:** two runs sharing a `model_artifacts_dir` overwrite each other's checkpoints with
no error. For anyone training a group of models, the natural setup (one artifacts directory, several
runs) silently yields a directory containing only the last run's weights, at every epoch number. The
result looks like a complete set of checkpoints.

**Fix:**

```python
model_path = f"{self.training_config.model_artifacts_dir}/{run_name}_epoch_{epoch}.pth"
```

The GCS path at line 451 has the same problem and needs the same change.

**Test:**

```python
def test_two_runs_do_not_overwrite_each_other():
    pipeline.save_model(epoch=0, run_name="run_a")
    pipeline.save_model(epoch=0, run_name="run_b")
    assert len(list(Path(artifacts_dir).glob("*_epoch_0.pth"))) == 2
```

---

# 6. `WandbWrapper` raises `KeyError` instead of degrading

**Where:** `bridge/infra/clients/wandb/wandb_wrapper.py:32`

**Now:**

```python
wandb.login(key=os.environ["WANDB_API_KEY"])
```

**The problem:** a bare subscript. If the variable is unset, this is an uncaught `KeyError`, not a
clear message about a missing credential.

**What it causes:** any user who enables metrics without a W&B account gets a `KeyError:
'WANDB_API_KEY'` from inside a logging wrapper, which does not suggest the actual remedy. Downstream
consumers work around it by writing a placeholder value into the environment purely to stop the
subscript raising.

**Fix:**

```python
api_key = os.environ.get("WANDB_API_KEY")
if not api_key:
    logger.warning("WANDB_API_KEY is not set; disabling W&B logging for this run.")
    self.is_enabled = False
    return
wandb.login(key=api_key)
```

---

# 7. `DeviceManager` exposes no way to change the device

**Where:** `bridge/utils/device_manager.py`. The singleton is constructed at import time on line 93
as `DeviceManager(device="cpu")`. `device` is a read-only property at lines 46-49. There is no setter,
no environment variable, and no configuration path.

**The problem:** the process-wide device is fixed to CPU at import, and every consumer snapshots
`device_manager.device` in its own `__init__`. There is no supported way to select a GPU.

**What it causes:** the only working override is to assign the private attribute
(`device_manager._device = torch.device("cuda:0")`) before constructing any BRIDGE object. Because
each object reads the device at construction time, an override applied even slightly too late is a
silent no-op that leaves everything on CPU. The symptom is a training run that is roughly an order of
magnitude slower than expected, with no error, on a machine with a perfectly good GPU.

**Fix:** add a setter and honour an environment variable.

```python
def set_device(self, device: str | torch.device) -> torch.device:
    """Repoint the manager. Objects constructed earlier keep the old device."""
    self.__init__(device=device)
    return self._device

device_manager = DeviceManager(device=os.environ.get("BRIDGE_DEVICE", "cpu"))
```

Raising rather than falling back when a device is named explicitly would be a further improvement.
Today `DeviceManager(device="cuda")` on a machine without CUDA logs a warning and silently returns
CPU, which is the same slow-run failure with a line in the log nobody reads.

---

# 8. `TrainingConfig` validation writes to the filesystem

**Where:** `bridge/domain/datamodels/training_config.py:34-36`

**Now:**

```python
values["model_artifacts_dir"] = os.path.join(project_root, values["model_artifacts_dir"])
os.makedirs(values["model_artifacts_dir"], exist_ok=True)
```

**The problem:** two things. Constructing a config object creates a directory as a side effect, so
validation is not pure. And a relative `model_artifacts_dir` is joined onto `get_project_root()`,
which is BRIDGE's own install root, not the caller's working directory.

**What it causes:** a user passing `model_artifacts_dir="runs/experiment1"` gets checkpoints written
inside the installed package, where nobody looks and where the next `uv sync` or `pip install -U`
deletes them. Absolute paths happen to work because `os.path.join` ignores the prefix when the second
argument is absolute, so the behaviour differs based on a property of the string that is not
documented. Additionally, merely validating a config in a test creates a directory.

**Fix:** resolve relative paths against the current working directory, and move directory creation to
the point of first write in `save_model`.

```python
p = Path(values["model_artifacts_dir"])
values["model_artifacts_dir"] = str(p if p.is_absolute() else Path.cwd() / p)
```

If joining onto the project root is intentional, please document it on the field, because the current
behaviour is indistinguishable from a bug at the call site.

---

# 9. `TrainingPipeline` is not constructible from the public API

**Where:** `bridge/__init__.py:25` (`__all__`) and
`bridge/infra/metrics/metrics_logger.py` (`metrics_logger_factory`).

**The problem:** `TrainingPipeline.__init__` takes a metrics logger positionally, and the only way to
build one is `metrics_logger_factory`, which is not in `bridge.__all__` and not re-exported. Assembling
a pipeline requires importing from `bridge.infra.metrics.metrics_logger` directly.

**What it causes:** every consumer reaches into the internal module layout, so any future
reorganization of `bridge.infra` is a breaking change for code that had no supported alternative.

**Fix:** add `metrics_logger_factory` and `MetricsConfig` to `bridge/__init__.py`'s `__all__`, or
default the parameter so a pipeline can be built without one:

```python
def __init__(self, model, training_config, dataset, metrics_logger=None):
    if metrics_logger is None:
        metrics_logger = metrics_logger_factory(MetricsConfig(...), training_config)
```

---

## Suggested order

1, 2, 4 and 5 are one-line or two-line changes with clear tests and no design questions. 3 needs a
config field. 6, 7, 8 and 9 involve small API decisions.

Defect 1 is the most urgent: any phonological accuracy number BRIDGE has reported to date is
overstated by an unknown, batch-dependent amount.

## Reproduction environment

`origin/main` at `5d7e8d9`, Python 3.12, `numpy<2`. The empirical check under defect 1 needs only
`BridgeTokenizer` and no trained model.

**Comment by mathnathan** [comment]
All nine defects are fixed and on `main` via #230 and #232.

This stayed open by mistake: #230's body said `Closes #223, #224, #225, #226, #227`, and
GitHub only parses a closing keyword attached to a single reference, so it closed #223 and
treated the rest as plain mentions.

| # | defect | commit | evidence |
|---|---|---|---|
| 1 | phon metrics mask padding with the orthographic pad id | `130f2b4` | see #227 |
| 2 | `seed = 0` silently trains unseeded | `508f492` | max parameter deviation between two `seed=0` models 5.146e+00 -> 0.000e+00; `seed=1` unchanged as the control |
| 3 | `TrainingPipeline` never shuffles | `ba05b06` | the pipeline now reorders the train partition per epoch, validation tail untouched |
| 4 | resume is dead code | `e593b4e` | see #226 |
| 5 | `save_model` accepts `run_name` and ignores it | `522aeef` | resolved by moving the policy out rather than picking a naming scheme, see below |
| 6 | `WandbWrapper` raises `KeyError` instead of degrading | `bdee203` | unset or empty `WANDB_API_KEY` now warns and disables |
| 7 | `DeviceManager` exposes no way to change the device | `e57d900` | `set_device` added, `BRIDGE_DEVICE` honoured; also fixed #223 |
| 8 | `TrainingConfig` validation writes to the filesystem | `a538ac0` | a relative `model_artifacts_dir` resolved under BRIDGE's *install root* and was created as a side effect; it now resolves against cwd and nothing is created until first write |
| 9 | `TrainingPipeline` not constructible from the public API | `86a09ec` | `metrics_logger_factory` exported; a test builds a full pipeline using only `from bridge import ...` |

Two notes on how the fixes differ from what this issue proposed.

**Defect 3** needed more than the sketch here. `TrainingConfig` had no `seed` field, so both
`seed` and `shuffle_each_epoch` were added. More importantly, the obvious test does not work:
counting words that have not moved passes for `shuffle(batch_size_train)` *and* for shuffling
within each batch. Both wrong fixes were planted and passed all eight tests. Batch membership
is what separates them from a real permutation, since it decides which words share a gradient
step, and that is what the test asserts now.

**Defect 5** was not fixed as prescribed. Putting the run name in the filename means the
library picks a naming scheme, and no scheme is right for every experiment now that BRIDGE is
imported and driven from experiment code. Instead `save_checkpoint(path, epoch)` replaces
`save_model(epoch, run_name)`, `save_every` is deleted, and `run_train_val_loop` yields a
`TrainingEvent` per step and writes nothing. The caller decides when and where to checkpoint.
That is a **breaking change**; the migration is in `docs/decisions/0006-the-caller-owns-the-training-loop.md`.

Defect 1's consequence is worth repeating here because it affects results rather than
ergonomics: every phonological metric BRIDGE reported before `130f2b4` is void and cannot be
corrected after the fact. Details in #227.
